pool/test/filter_assignment_tests.ml

```diff
@@ -0,0 +1,335 @@
+  let* verification_events =
+    let open Cqrs_command.User_command in
+    let created_email =
+      Email.Created (email, token, user_id) |> Pool_event.email_verification
+    in
+    let email = Email.create email contact.user token in
+    let@ verify_events = VerifyEmail.handle (Contact contact) email in
+    Ok (created_email :: verify_events)
+  in
+  let& () =
+    Pool_event.handle_events test_db verification_events |> Lwt_result.ok
+  in
+  let& contact = Contact.find test_db user_id in
+  let verification_events =
+    [ Contact.Verified contact |> Pool_event.contact ]
+  in
+  let& () =
+    Pool_event.handle_events test_db verification_events |> Lwt_result.ok
+  in
+  let& contact = Contact.find test_db user_id in
+  Lwt_result.lift (Ok contact)
+;;
+
+let assignment ~experiment ~session ~contact =
+  let open Cqrs_command.Assignment_command in
+  let already_enrolled = false in
+  let* events =
+    Create.(
+      handle
+        { experiment; contact; follow_up_sessions = []; session }
+        (fun _ ->
+          Sihl_email.create
+            ~sender:"sender"
+            ~recipient:"recipient"
+            ~subject:"subject"
+            "body")
+        already_enrolled)
+  in
+  let& () = Pool_event.handle_events test_db events |> Lwt_result.ok in
+  Lwt_result.lift (Ok ())
+;;
+
+let invitation ~experiment ~contacts =
+  let open Cqrs_command.Invitation_command in
+  let* events =
+    Create.(
+      handle
+        { experiment
+        ; contacts
+        ; invited_contacts = []
+        ; create_message =
+            (fun _ ->
+              Sihl_email.create
+                ~sender:"sender"
+                ~recipient:"recipient"
+                ~subject:"subject"
+                "body"
+              |> Result.ok)
+        ; mailing = None
+        })
+  in
+  let& () = Pool_event.handle_events test_db events |> Lwt_result.ok in
+  Lwt_result.lift (Ok ())
+;;
+
+(** This test verifies that given a contact that has accpeted an invitation to an experiment,
+    and one that was not, the contact that did not accept will show after the
+    assignment exclusion filter is applied.
+
+    It does so by:
+
+    1. creating an experiment
+    2. creating a contact that has accepted an invitation to the experiment
+    3. creating a contact that has NOT accepted an invitation to the experiment
+    4. create a filter that for assignments that includes our experiment
+    5. assert on the found contacts
+
+    Fin. *)
+let finds_unassigned_contacts =
+  Test_utils.case
+  @@ fun () ->
+  (* 1. creating an experiment *)
+  let& experiment = experiment () in
+  (* 2. creating an session *)
+  let& session = session ~experiment in
+  (* 2. creating contacts *)
+  let& assigned_contact = contact ~prefix:"invited" () in
+  let& unassigned_contact = contact ~prefix:"probe" () in
+  (* 3. send invitations *)
+  let& _invitation =
+    invitation ~experiment ~contacts:[ assigned_contact; unassigned_contact ]
+  in
```

Comment:
@mabiede i still create an invitation here, but in the system itself is technically not needed. As long as there's an assignment (created in the immediate next let) this will work.

Same for [line 287](https://github.com/uzh/pool/pull/267/files#diff-12fd6fdcb7f5643497a302117d8a81dc2fef08db84d3e6fce99e7d9fa64bf973R287)

Reply:
Thats fine, as `_invitation` and `_assignment` are both of type unit, they should be updated like the following:
 
```suggestion
  let& () = invitation ~experiment ~contacts:[ assigned_contact; unassigned_contact ] in
```

Otherwise we prefer the type declaration for this kind of variables, so e.g. `let& (_ : Invitation.t) = ...`